classes/controllers/tab/workflow/PKPVersioningTabHandler.inc.php

```diff
@@ -0,0 +1,153 @@
+
+		// get data of the old version
+		$submission = $submissionDao->getById($submissionId, null, false, $oldVersion);
+		$authorDao = DAORegistry::getDAO('AuthorDAO');
+		$authors = $authorDao->getBySubmissionId($submissionId, true, false, $oldVersion);
+
+		// save new submission version without publication date
+		$newVersion = $oldVersion+1;
+		$submission->setData('submissionRevision', $newVersion);
+		$submission->setDatePublished(null);
+		$submissionDao->updateObject($submission);
+
+		// copy the authors from old version to new version
+		foreach($authors as $author) {
+			$authorId = (int)$author->getId();
+			$author->setVersion($newVersion);
+			$authorDao->insertObject($author, true);
+			$newAuthorId = (int)$authorDao->getInsertId();
+		}
+
+		// reload page to display new version
+		$dispatcher = $this->getDispatcher();
+		$redirectUrl = $dispatcher->url($request, ROUTE_PAGE, null, 'workflow', 'index', array($submission->getId(), $submission->getStageId()));
+		return $request->redirectUrlJson($redirectUrl);
+	}
+
+	/**
+	* Make a copy of the file to the specified file stage
+	* @param $context Context
+	* @param $submissionFile SubmissionFile
+	* @param $fileStage int SUBMISSION_FILE_...
+	* @return newFileId int
+	*/
+	function copyFile($context, $submissionFile, $fileStage){
+		import('lib.pkp.classes.file.SubmissionFileManager');
+		$submissionFileManager = new SubmissionFileManager($context->getId(), $submissionFile->getSubmissionId());
+		$fileId = $submissionFile->getFileId();
+		$revision = $submissionFile->getRevision();
+		list($newFileId, $newRevision) = $submissionFileManager->copyFileToFileStage($fileId, $revision, $fileStage, null, true);
+		return $newFileId;
+	}
+
+	/**
+	 * @see PKPHandler::setupTemplate
+	 */
+	function setupTemplate($request) {
+		AppLocale::requireComponents(LOCALE_COMPONENT_APP_EDITOR);
+		parent::setupTemplate($request);
+	}
+
+	/**
+	 * Handle version info (tab content).
+	 * @param $request PKPRequest
+	 * @param $args array
+	 * @return JSONMessage JSON object
+	 */
+	function versioning($args, $request) {
+		$this->setupTemplate($request);
+		$templateMgr = TemplateManager::getManager($request);
+
+		// Retrieve the authorized submission, stage id and submission revision.
+		$submission = $this->getAuthorizedContextObject(ASSOC_TYPE_SUBMISSION);
+		$stageId = $this->getAuthorizedContextObject(ASSOC_TYPE_WORKFLOW_STAGE);
+		$submissionRevision = $this->getAuthorizedContextObject(ASSOC_TYPE_SUBMISSION_REVISION);
+		
+		// Add variables to the template.
+		$templateMgr = TemplateManager::getManager($request);
+		$templateMgr->assign('submission', $submission);
+		$templateMgr->assign('stageId', $stageId);
+		$templateMgr->assign('submissionRevision', $submissionRevision);
+
+		// Get publication date of submission revision 
+		$submissionDao = Application::getSubmissionDAO();
+		$submission = $submissionDao->getById($submission->getId(), null, false, $submissionRevision);
```

Comment:
hmm, I thought we would need to get the object again with the submission revision to get the correct date. But it seems to work anyway.... 😣 👍